pypck/helpers.py

```diff
@@ -0,0 +1,33 @@
+"""Helper functions for pypck."""
+
+import asyncio
+from typing import Any, Awaitable, List
+
+PYPCK_TASKS: List["asyncio.Task[Any]"] = []
+
+
+def create_task(coro: Awaitable[Any]) -> "asyncio.Task[None]":
+    """Create a task and store a reference in the task registry."""
+    task = asyncio.create_task(coro)
+    task.add_done_callback(PYPCK_TASKS.remove)
+    PYPCK_TASKS.append(task)
+    return task
+
+
+async def cancel_task(task: "asyncio.Task[Any]") -> bool:
+    """Cancel a task.
+
+    Wait for cancellation completed but do not propagate a possible CancelledError.
+    """
+    success = task.cancel()
+    try:
+        await task
+    except asyncio.CancelledError:
+        pass
+    return success  # was not already done
+
+
+async def cancel_all_tasks() -> None:
+    """Cancel all pypck tasks."""
+    for task in tuple(PYPCK_TASKS):
+        await cancel_task(task)
```

Comment:
@alengwenus I just came across this PR.  And I very much like it since it does away with the tedious babysitting of fire and forget tasks. :+1: 

Just a quick question: I very much understand that you copy and iterate the tasks list in order to avoid "modify while iterate" issues. However, as far as I can see it's quite possible that a task in the list adds another tasks while the cancel loop is in progress.

I guess it's deliberately considered ok? Or should it rather make sure that when the function returns, the list is guaranteed to be empty?

Reply:
Well ok, that could happen. But how should we solve it? What about an is_cancelling flag that prevents create_task from spawning new tasks?
Any idea?

Reply:
Either that. Or something like

```py
while len(PYPCK_TASKS) > 0:
    task = PYPCK_TASKS[0]
    await cancel_task(task)
```

However, this requires that by the time `cancel_task` returns, the task is removed by the done callback. Not sure if python guarantees that.  If it's not guaranteed, I'd have to think a bit about the issue.

And there is always the option to just punt and add a comment that says that there might be tasks in the list after cancel if they where added after the function was started. :shrug:

Reply:
Last time I checked, I didn't find anything in the official docs about when the done callback is invoked.

Reply:
It's documented [here](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.add_done_callback), respectively [here](https://docs.python.org/3/library/asyncio-future.html#asyncio.Future.add_done_callback).
And [here](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task.done) is something about when a task turns to *done*. So either the task has to return a value, raise an exception, or has been cancelled.

Reply:
The following test program
```
import asyncio

async def test():
    task = asyncio.create_task(asyncio.sleep(60));
    task.add_done_callback(lambda task: print('callback'));
    task.cancel()
    try:
        await task
    except:
        pass
    print('awaited')

asyncio.run(test())
```
prints 
```
callback
awaited
```
for me. So it appears consistent with your and my interpretation of the reference docs. So I guess the solution should be fine.

Reply:
If you want to be extra cautions, we could put an assertion (or alternatively a warning) if two consecutive iterations of the while loop would see the same task object.  But I'm not sure it's worth the extra lines of code.